cranelift/wasm/src/code_translator.rs

```diff
@@ -2142,6 +2144,42 @@ fn translate_icmp(cc: IntCC, builder: &mut FunctionBuilder, state: &mut FuncTran
     state.push1(builder.ins().bint(I32, val));
 }
 
+fn fold_atomic_mem_addr(
+    linear_mem_addr: Value,
+    memarg: &MemoryImmediate,
+    access_ty: Type,
+    builder: &mut FunctionBuilder,
+) -> Value {
+    let access_ty_bytes = access_ty.bytes();
+    let final_lma = if memarg.offset > 0 {
+        assert!(builder.func.dfg.value_type(linear_mem_addr) == I32);
+        let linear_mem_addr = builder.ins().uextend(I64, linear_mem_addr);
+        let a = builder
+            .ins()
+            .iadd_imm(linear_mem_addr, i64::from(memarg.offset));
+        let cflags = builder.ins().ifcmp_imm(a, 0x1_0000_0000i64);
+        builder.ins().trapif(
```

Comment:
I figured that we already had a function somewhere in this file to calculate an address, could that perhaps be used instead of duplicating the logic here? That handles things with the guard page and whatnot too to deduplicate checks.

Reply:
In Cranelift memory instructions directlt get an offset as argument, so there is probably no need for such a function.

Reply:
I could not find specifically one that combines offsets, I saw only those that calculate a pointer based on memory base.

Reply:
@bjorn3 But for atomic wait/notify, the actual handling code is generated by calling `environ.translate_atomic_wait` etc.  And that takes only the final linear memory offset as an argument.  There's no way to pass it an additional constant offset as a second arg.  Hence we have no option here but to "manually" fold in `memarg.offset` (and adding a second bounds check, sigh.)

Reply:
Indeed. I was reacting to

> I figured that we already had a function somewhere in this file to calculate an address

Reply:
Could raw pointers get passed to the native functions which handle the implementation? The bounds-checking code is so tricky it seems like it'd be best to avoid duplicating it if we can.

Reply:
Yep, raised the related question at https://bugzilla.mozilla.org/show_bug.cgi?id=1684861#c7 (consumer side)